source/common/router/upstream_request.cc

```diff
@@ -371,9 +378,25 @@ void UpstreamRequest::onPoolReady(
   // If end_stream is set in headers, and there are metadata to send, delays end_stream. The case
   // only happens when decoding headers filters return ContinueAndEndStream.
   const bool delay_headers_end_stream = end_stream && !downstream_metadata_map_vector_.empty();
+  // Make sure that when we are forwarding CONNECT payload we do not do so until
+  // the upstream has accepted the CONNECT request.
+  if (conn_pool_->protocol().has_value() && headers->Method() &&
+      headers->Method()->value().getStringView() == Http::Headers::get().MethodValues.Connect) {
+    paused_for_connect_ = true;
+  }
+
   upstream_->encodeHeaders(*parent_.downstreamHeaders(), end_stream && !delay_headers_end_stream);
   calling_encode_headers_ = false;
 
+  if (!paused_for_connect_) {
+    encodeBodyAndTrailers();
+  }
+}
+
+void UpstreamRequest::encodeBodyAndTrailers() {
+  const bool end_stream = !buffered_request_body_ && encode_complete_ && !encode_trailers_;
+  const bool delay_headers_end_stream = end_stream && !downstream_metadata_map_vector_.empty();
```

Comment:
nit: this is pretty specific/confusing/copied from above. Small helper that returns a struct/tuple/something with comments that can be shared with above?